### PR TITLE
MyMiniFactory Driver Feature

### PR DESCRIPTION
**Pull request by BohunkG4mer**
Here is the first draft of the MyMiniFactory driver feature. I know there is a message feature you added to the main that I still need to integrate into this pull but I figured I'd put this up here for you to review and give feedback at your leisure while I look into that and adding a preference option for which service to start with.

### CODE REVIEW

**Comment by ChrisTerBeke** [comment]
Nice! I'll look through this coming days.

File: thingiverse/ThingiverseService.py [hunk 1]
```diff
@@ -41,31 +48,45 @@ class ThingiverseService(QObject):
...
-        CuraApplication.getInstance().getPreferences().preferenceChanged.connect(self._onPreferencesChanged)
-        self._user_name = self._initSettings(Settings.SETTINGS_USER_NAME_PREFERENCES_KEY)
+        self._initSettings(Settings.THINGIVERSE_USER_NAME_PREFERENCES_KEY, "")
+        self._initSettings(Settings.MYMINIFACTORY_USER_NAME_PREFERENCES_KEY, "")
 
         # List of supported file types.
         self._supported_file_types = []  # type: List[str]
 
+    @pyqtSlot(str, name="setService")
+    def setService(self, value: str) -> None:
```
Comment (ChrisTerBeke) [inline thingiverse/ThingiverseService.py]: Instead of this we could keep a dict of 'providers' and call the right function on the right provider depending on the actively selected service.

File: thingiverse/ThingiverseService.py [hunk 2]
```diff
@@ -78,13 +99,29 @@ def openSettings(self) -> None:
...
+    @pyqtProperty(list, notify=apiClientChanged)
+    def disabledViews(self) -> list:
+        """
+        Get any disabled views for the current API Client
+        :return: List of view labels that should not be visible
+        """
+        return self._api_client.disabled_views
+
+    @pyqtProperty(str, notify=userNameChanged)
+    def thingiverseUserName(self) -> str:
```
Comment (ChrisTerBeke) [inline thingiverse/ThingiverseService.py]: Somehow I'd like to unify these two username methods.

File: thingiverse/ThingiverseService.py [hunk 3]
```diff
@@ -123,15 +161,15 @@ def isQuerying(self) -> bool:
         return self._is_querying
 
     @pyqtProperty("QVariant", notify=activeThingChanged)
-    def activeThing(self) -> Dict[str, any]:
+    def activeThing(self) -> Thing:
```
Comment (ChrisTerBeke) [inline thingiverse/ThingiverseService.py]: This type actually didn't change, as the returned data is a dict representation of the object (otherwise Qt craps out).

File: thingiverse/ThingiverseService.py [hunk 4]
```diff
@@ -123,15 +161,15 @@ def isQuerying(self) -> bool:
...
         """
         Get the current active thing details.
         :return: The thing.
         """
         return self._thing_details.__dict__ if self._thing_details else None
 
     @pyqtProperty("QVariantList", notify=activeThingFilesChanged)
-    def activeThingFiles(self) -> List[Dict[str, any]]:
```
Comment (ChrisTerBeke) [inline thingiverse/ThingiverseService.py]: Same as above.

File: thingiverse/ThingiverseService.py [hunk 5]
```diff
@@ -169,7 +207,9 @@ def getLiked(self) -> None:
         """
         if not self._checkUserNameConfigured():
             return
-        self._executeQuery("users/{}/likes".format(self._user_name))
+        self._clearSearchResults()
+        query = self._api_client.getLikesUrl()
+        self._executeQuery(query)
```
Comment (ChrisTerBeke) [inline thingiverse/ThingiverseService.py]: Better to have this _clearSearchResults and is_from_collection logic in self._executeQuery? Now its in multiple places, making it hard to keep behavior up to date

File: thingiverse/ThingiverseService.py (hunk 2 above)
Comment (ChrisTerBeke) [reply thingiverse/ThingiverseService.py]: Maybe we should move it to the ApiClient implementations (let them store their own username) and then just call the right client here (just like we do with all the other methods). That would keep all the service-specific things together.

File: thingiverse/ThingiverseService.py [hunk 6]
```diff
@@ -316,14 +377,14 @@ def _onDownloadFinished(self, file_bytes: bytes, file_name: str) -> None:
         self._is_downloading = False
         self.downloadingStateChanged.emit()
 
-    def _onQueryFinished(self, things: List[JSONObject]) -> None:
+    def _onQueryFinished(self, objects: List[JSONObject]) -> None:
```
Comment (ChrisTerBeke) [inline thingiverse/ThingiverseService.py]: Not sure if objects is a better name, it's quite generic (in relation to Python).

File: thingiverse/ThingiverseService.py [hunk 7]
```diff
@@ -299,8 +358,10 @@ def _onThingFilesFinished(self, thing_files: List[JSONObject]) -> None:
         Callback for receiving a list of thing files on. Filtered on supported file types of Cura.
         :param thing_files: The thing files.
         """
-        self._thing_files = [f for f in thing_files if pathlib.Path(f.name).suffix.lower().strip(".")
-                             in self._supported_file_types]
+        self._thing_files = []
+        for f in thing_files:
```
Comment (ChrisTerBeke) [inline thingiverse/ThingiverseService.py]: Why this change?

**Comment by ChrisTerBeke** [comment]
Btw, I might pick some of these comments up myself if I feel like coding.

**Comment by BohunkG4mer** [comment]
I will work on adding in a preference for the ApiClient to use and then check in with you on what I should do next. Thanks!

**Comment by ChrisTerBeke** [comment]
@BohunkG4mer I'm currently working on a branch based on your work (while processing my review comments and doing some more refactoring). I'll push this ASAP, probably it's best to base your further work on that one?

File: api/APIClient.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+        request = QNetworkRequest(QUrl().fromUserInput(url))
+        request.setHeader(QNetworkRequest.ContentTypeHeader, content_type)
+        request.setAttribute(QNetworkRequest.RedirectPolicyAttribute, True)  # file downloads reply with a 302 first
+        auth_type, auth = self._auth
+        if auth is not None:
+            if auth_type == "Authorization":
+                request.setRawHeader(b"Authorization", auth)
+            else:
+                operator = "&" if url.find("?") > 0 else "?"
+                request.setUrl(QUrl.fromUserInput("{}{}{}".format(url, operator, auth)))
```
[the remark this answers is not in the document]
Comment (BohunkG4mer) [reply api/APIClient.py]: Yes, MyMiniFactory does authentication by append a "key" url get parameter with the public API key for the app that I created there. I had trouble using their OAuth impmentation extending the OAuth class in Cura so I punted and did this instead.

File: api/APIClient.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+                if not on_failed:
+                    return
+                return on_failed(None)
+            on_finished(response)
+
+        self._anti_gc_callbacks.append(parse)
+        reply.finished.connect(parse)
+
+    @staticmethod
+    def _jsonDecoder(data: dict) -> JSONObject:
```
[the remark this answers is not in the document]
Comment (BohunkG4mer) [reply api/APIClient.py]: From my understanding, this function, which provided as a hook for json.loads only runs per object so it shouldn't ever receive anything other that a dict

File: api/JSONObject.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+class JSONObject(QObject):
+    """ Simple class that converts a JSON object to a Python model. """
+    def __init__(self, _dict: Dict[str, Any]):
+        self.TYPE = self.__class__.__name__
+        if _dict:
+            vars(self).update(_dict)
+        super().__init__()
+
+class BaseModel(JSONObject):
+    URL = None # type: str
```
[the remark this answers is not in the document]
Comment (BohunkG4mer) [reply api/JSONObject.py]: To not be confused with the provided dict values I guess. No real reason they have to be

File: myminifactory/MyMiniFactoryApiClient.py [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+        return ["My Makes"]
+
+    # API constants.
+    @property
+    def _root_url(self):
+        return "https://www.myminifactory.com/api/v2"
+
+    @property
+    def _auth(self):
+        return "Query Key", "key={}".format(Settings.MYMINIFACTORY_API_TOKEN)
```
[the remark this answers is not in the document]
Comment (BohunkG4mer) [reply myminifactory/MyMiniFactoryApiClient.py]: It only supports it in the query. Regardless if it is in the url or header, clients have access to see it and both are the same amount of secure. MyMiniFactory support said it's secure and just to track who is making requests to specific urls. I would like to switch to OAuth authentication but I ran into issues and gave up. Maybe I will look back into that some time here.

**Comment by BohunkG4mer** [comment]
Sounds good. I will adjust my upstream or just wipe and set up a new one then. Thanks for all the feedback and pointers. My python experience is skyrocketing! :)

**Comment by ChrisTerBeke** [comment]
> Sounds good. I will adjust my upstream or just wipe and set up a new one then. Thanks for all the feedback and pointers. My python experience is skyrocketing! :)

Great to hear you're improving your skills this way! I think making PRs on open source projects and getting feedback is a very nice way to learn, and it helps the authors of the projects a lot!

File: myminifactory/MyMiniFactoryApiClient.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (BohunkG4mer) [reply myminifactory/MyMiniFactoryApiClient.py]: I will double check that it can't be sent in a header though

File: api/APIClient.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ChrisTerBeke) [reply api/APIClient.py]: Is the current key a static key? Or it is manually obtained via an OAuth2 flow?

**Comment by ChrisTerBeke** [comment]
Closing in favour of #19.

File: thingiverse/ThingiverseService.py (hunk 3 above)
Comment (nallath) [reply thingiverse/ThingiverseService.py]: You can get a dict up & running with "QVariantMap"
